Add HTTP server

theacodes opened the pull request.

Commit:
Add HTTP server.

dhermes commented on Procfile:
```diff
@@ -1 +1,2 @@
 grpc: python shiny/grpc_server.py
+http: gunicorn -b :8080 shiny.http_server:app
```
How appropriate that we are using `gunicorn`. `-b` --> `--bind`

dhermes commented on shiny/http_server.py:
```diff
@@ -0,0 +1,13 @@
+
+@app.route('/v1/do-nothing/<unicorn_name>', methods=['POST'])
+def do_nothing(unicorn_name):
+    return ''
```
This is cheating. Can we at least grab the path from the `.proto` file somehow? At the very least it is in `shiny_pb2.DESCRIPTOR.serialized_pb`, but in an opaque way.

Also, prefer this:

``` python
from google.protobuf import empty_pb2
from google.protobuf import json_format

def do_nothing(unicorn_name):
    result = empty_pb2.Empty()
    return json_format.MessageToDict(result)
```

dhermes replied:
OK I figured it out.

``` python
>>> from google.protobuf import descriptor_pb2
>>>
>>> from shiny import shiny_pb2
>>>
>>> file_pb = descriptor_pb2.FileDescriptorProto.FromString(
...     shiny_pb2.DESCRIPTOR.serialized_pb)
>>> service, = file_pb.service
>>> service.name
u'Shiny'
>>> method_pb, = service.method
>>> (field, http_rule), = method_pb.options.ListFields()
>>> field.full_name
'google.api.http'
>>> http_rule.post
u'/v1/do-nothing/{unicorn.name}'
```

dhermes replied:
It seems Pub/Sub returns an `Empty` on [some POST requests](https://github.com/googleapis/googleapis/blob/5cf7f5a30210bb5750d6918b30b288da2b2b91ec/google/pubsub/v1/pubsub.proto#L67-L80) so [I checked](https://gist.github.com/dhermes/3e94adb65ea0e11ba664b68e1b4d72d7) what the JSON-over-HTTP interface returns and confirmed it is an empty dictionary.

theacodes replied:
Do you want me to still do this? It'll take some massaging of the path, but it's possible.

dhermes replied:
It's definitely worth a try.

theacodes replied:
I feel like I'm slowly re-inventing endpoints v1.

dhermes replied:
In function maybe, but in form, not even close (recall I launched that API).

theacodes replied:
This is getting way too complicated. Let's just table it until endpoints can do the translation for us.

dhermes committed:
HTTP server/client updates.

- On the server, actually construct the PB request object
- On the client, actually send the "transmogrify" value

dhermes commented:
Yo @jonparrott I just pushed a commit into this branch. LMK what you think. If you don't like it, just `git reset --force HEAD^` and force push over it?

dhermes commented on shiny/http_server.py:
```diff
@@ -0,0 +1,30 @@
+@app.route('/v1/do-nothing/<unicorn_name>', methods=['POST'])
+def do_nothing(unicorn_name):
+    combined = copy.deepcopy(flask.request.json)
+    combined.update(flask.request.args.to_dict())
```
@jonparrott It's worth noting here that `flask.request.args` is a multidict, so calling `to_dict()` may clobber some values, e.g.

``` python
>>> from werkzeug.datastructures import ImmutableMultiDict
>>> d = ImmutableMultiDict([
...     ('a', 'b'),
...     ('c', 'd'),
...     ('e', 'f'),
...     ('a', 'pickle'),
...     ('c', 'soup'),
... ])
>>> d.to_dict()
{'a': 'b', 'c': 'd', 'e': 'f'}
```

Commit:
Slight re-org

theacodes commented:
I factored out the request to proto part, PTAL. If you're cool with it, merge it.
